### PR TITLE
added CarAuditLogItemDao + impl

### PR DESCRIPTION
**Pull request by rtrembecky**
- fixed Car method name getID() -> getId()

### CODE REVIEW

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public List<CarAuditLogItem> findAll() {
+        return em.createQuery(
+                "SELECT i FROM CarAuditLogItem i", CarAuditLogItem.class).getResultList();
+    }
+
+    @Override
+    public List<CarAuditLogItem> findById(UUID id) {
+        TypedQuery<CarAuditLogItem> query = em.createQuery(
+                "SELECT i FROM CarAuditLogItem i WHERE i = :itemId", CarAuditLogItem.class);
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: should be ` WHERE i.id = :itemId`

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    private EntityManager em;
+
+    @Override
+    public List<CarAuditLogItem> findAll() {
+        return em.createQuery(
+                "SELECT i FROM CarAuditLogItem i", CarAuditLogItem.class).getResultList();
+    }
+
+    @Override
+    public List<CarAuditLogItem> findById(UUID id) {
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: Returns one log item or NULL.

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        TypedQuery<CarAuditLogItem> query = em.createQuery(
+                "SELECT i FROM CarAuditLogItem i WHERE i = :itemId", CarAuditLogItem.class);
+
+        query.setParameter("itemId", id);
+        return query.getResultList();
+    }
+
+    @Override
+    public List<CarAuditLogItem> findByCar(Car c) {
+        TypedQuery<CarAuditLogItem> query = em.createQuery(
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: better, right?

```java
TypedQuery<CarAuditLogItem> query = em.createQuery(
                "SELECT i FROM CarAuditLogItem i WHERE i.car = :car", CarAuditLogItem.class);
        query.setParameter("car", c);
```

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        query.setParameter("carId", c.getId());
+        return query.getResultList();
+    }
+
+    @Override
+    public List<CarAuditLogItem> findByUser(User u) {
+        TypedQuery<CarAuditLogItem> query = em.createQuery(
+                "SELECT i FROM CarAuditLogItem i WHERE i.user = :userId", CarAuditLogItem.class);
+
+        query.setParameter("userId", u.getId());
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: same thing like in findByCar

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public List<CarAuditLogItem> getRecordsCreatedBetween(Date from, Date to) {
+        TypedQuery<CarAuditLogItem> query = em.createQuery(
+                "SELECT r FROM CarAuditLogItem r WHERE r.created BETWEEN :fromDate AND :toDate ", CarAuditLogItem.class);
+
+        query.setParameter("fromDate", from);
+        query.setParameter("toDate", to);
+        return query.getResultList();
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: Great!!!

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public void create(CarAuditLogItem i) {
+        em.persist(i);
+    }
+
+    @Override
+    public void delete(CarAuditLogItem i) {
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: create() and delete() not possible, due the fact that `CarAuditLogItem` is the abstract class.

File: src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java (hunk 1 above)
Comment (rtrembecky) [inline src/main/java/cz/fi/muni/pa165/Model/Dao/CarAuditLogItemDaoImpl.java]: should be fixed in next commit, along the others

**Comment by charlliz** [comment]
In this case, when we want to use abstract class in EntityManager plese mark class with annotation @Entity instead of @MappedSuperclass (this is my fault I marked it before when I thought it will not be used)

**Comment by jkuchar** [comment]
@charlliz @rtrembecky 👌👌🏻👌🏼👌🏽👌🏾👌🏿

**Comment by rtrembecky** [comment]
cool now?

File: src/main/java/cz/fi/muni/pa165/Model/Entity/CarAuditLogItem.java [hunk 7]
```diff
@@ -6,10 +6,9 @@
 import javax.persistence.Entity;
 import javax.persistence.Id;
 import javax.persistence.ManyToOne;
-import javax.persistence.MappedSuperclass;
 import javax.validation.constraints.NotNull;
 
-@MappedSuperclass
```
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Entity/CarAuditLogItem.java]: are you sure? We will see in tests. :-)

Comment (charlliz) [inline src/main/java/cz/fi/muni/pa165/Model/Entity/CarAuditLogItem.java]: Yes I'm sure. http://docs.oracle.com/javaee/6/tutorial/doc/bnbqn.html

**Comment by charlliz** [comment]
I think it's ok now 👍

File: src/main/java/cz/fi/muni/pa165/Model/Entity/CarAuditLogItem.java (hunk 7 above)
Comment (jkuchar) [inline src/main/java/cz/fi/muni/pa165/Model/Entity/CarAuditLogItem.java]: Cool!

**Pull request merged**